pkg/nimble/netif/nimble_netif.c

```diff
@@ -0,0 +1,621 @@
+            *((gnrc_nettype_t *)value) = _nettype;
+            res = sizeof(gnrc_nettype_t);
+            break;
+        case NETOPT_DEVICE_TYPE:
+            assert(max_len == sizeof(uint16_t));
+            *((uint16_t *)value) = NETDEV_TYPE_BLE;
+            res = sizeof(uint16_t);
+            break;
+        default:
+            break;
+    }
+
+    return res;
+}
+
+static inline int _netdev_set(netdev_t *dev, netopt_t opt,
+                              const void *value, size_t val_len)
+{
+    (void)dev;
+    int res = -ENOTSUP;
+
+    switch (opt) {
+        case NETOPT_PROTO:
+            assert(val_len == sizeof(gnrc_nettype_t));
+            memcpy(&_nettype, value, sizeof(gnrc_nettype_t));
+            res = sizeof(gnrc_nettype_t);
+            break;
+        default:
+            break;
+    }
+
+    return res;
+}
+
+static const netdev_driver_t _nimble_netdev_driver = {
+    .send = NULL,
+    .recv = NULL,
+    .init = _netdev_init,
+    .isr  =  NULL,
+    .get  = _netdev_get,
+    .set  = _netdev_set,
+};
+
+static netdev_t _nimble_netdev_dummy = {
+    .driver = &_nimble_netdev_driver,
+};
+
+
+static int _on_data(nimble_netif_conn_t *conn, struct ble_l2cap_event *event)
+{
+    int ret = 0;
+    struct os_mbuf *rxb = event->receive.sdu_rx;
+    size_t rx_len = (size_t)OS_MBUF_PKTLEN(rxb);
+
+    /* allocate netif header */
+    gnrc_pktsnip_t *if_snip = gnrc_netif_hdr_build(conn->addr, BLE_ADDR_LEN,
+                                                   _nimble_netif->l2addr,
+                                                   BLE_ADDR_LEN);
+    if (if_snip == NULL) {
+        ret = NIMBLE_NETIF_NOMEM;
+        goto end;
+    }
+
+    /* we need to add the device PID to the netif header */
+    gnrc_netif_hdr_t *netif_hdr = (gnrc_netif_hdr_t *)if_snip->data;
+    netif_hdr->if_pid = _nimble_netif->pid;
+
+    /* allocate space in the pktbuf to store the packet */
+    gnrc_pktsnip_t *payload = gnrc_pktbuf_add(if_snip, NULL, rx_len, _nettype);
+    if (payload == NULL) {
+        gnrc_pktbuf_release(if_snip);
+        ret = NIMBLE_NETIF_NOMEM;
+        goto end;
+    }
+
+    /* copy payload from mbuf into pktbuffer */
+    int res = os_mbuf_copydata(rxb, 0, rx_len, payload->data);
+    if (res != 0) {
+        gnrc_pktbuf_release(payload);
+        ret = NIMBLE_NETIF_DEVERR;
+        goto end;
+    }
+
+    /* finally dispatch the receive packet to GNRC */
+    if (!gnrc_netapi_dispatch_receive(payload->type, GNRC_NETREG_DEMUX_CTX_ALL,
+                                      payload)) {
+        gnrc_pktbuf_release(payload);
+    }
+
+end:
+    /* copy the receive data and free the mbuf */
+    os_mbuf_free_chain(rxb);
+    /* free the mbuf and allocate a new one for receiving new data */
+    rxb = os_mbuf_get_pkthdr(&_mbuf_pool, 0);
+    /* due to buffer provisioning, there should always be enough space */
+    assert(rxb != NULL);
+    ble_l2cap_recv_ready(event->receive.chan, rxb);
+
+    return ret;
```

Comment:
If I see correctly the return value is never used. So is it needed?

Reply:
nope, its a leftover. Will remove

Reply:
done.